iothub/device/src/Transport/Mqtt/OrderedTwoPhaseWorkQueue.cs

```diff
@@ -15,71 +16,79 @@ private class IncompleteWorkItem
         {
             public IncompleteWorkItem(TWorkId id, TWork workItem)
             {
-                this.WorkItem = workItem;
-                this.Id = id;
+                WorkItem = workItem ?? throw new ArgumentNullException(nameof(workItem));
+                Id = id ?? throw new ArgumentNullException(nameof(id));
             }
 
             public TWork WorkItem { get; }
 
             public TWorkId Id { get; }
         }
 
-        private readonly Func<TWork, TWorkId> getWorkId;
-        private readonly Func<IChannelHandlerContext, TWork, Task> completeWork;
-        private readonly Queue<IncompleteWorkItem> incompleteQueue = new Queue<IncompleteWorkItem>();
+        private readonly Func<TWork, TWorkId> _getWorkId;
+        private readonly Func<IChannelHandlerContext, TWork, Task> _completeWork;
+        private readonly Queue<IncompleteWorkItem> _incompleteQueue = new Queue<IncompleteWorkItem>();
 
-        public OrderedTwoPhaseWorkQueue(Func<IChannelHandlerContext, TWork, Task> worker, Func<TWork, TWorkId> getWorkId, Func<IChannelHandlerContext, TWork, Task> completeWork)
+        public OrderedTwoPhaseWorkQueue(
+            Func<IChannelHandlerContext, TWork, Task> worker,
+            Func<TWork, TWorkId> getWorkId,
+            Func<IChannelHandlerContext, TWork, Task> completeWork)
             : base(worker)
         {
-            this.getWorkId = getWorkId;
-            this.completeWork = completeWork;
+            _getWorkId = getWorkId ?? throw new ArgumentNullException(nameof(getWorkId));
+            _completeWork = completeWork ?? throw new ArgumentNullException(nameof(completeWork));
         }
 
         public Task CompleteWorkAsync(IChannelHandlerContext context, TWorkId workId)
         {
-            if (this.incompleteQueue.Count == 0)
+            if (!_incompleteQueue.Any())
             {
                 throw new IotHubException("Nothing to complete.", isTransient: false);
             }
-            IncompleteWorkItem incompleteWorkItem = this.incompleteQueue.Peek();
-            if (incompleteWorkItem.Id.Equals(workId))
+
+            IncompleteWorkItem incompleteWorkItem = _incompleteQueue.Peek();
+            if (incompleteWorkItem != null
+                && incompleteWorkItem.Id.Equals(workId))
             {
-                this.incompleteQueue.Dequeue();
-                return this.completeWork(context, incompleteWorkItem.WorkItem);
+                _incompleteQueue.Dequeue();
```

Comment:
I don't love that we peek an item, validate it, and then assume when we call dequeue we get the same item. Perhaps that is a safe assumption, but I'd prefer our code not make such assumptions. I didn't change this, though - just noting I don't care for it.

Ideas?

Reply:
Either we lock around this and make it a transaction of sorts, or we dequeue, check and requeue. So that means either we steal work or we steal threads.

Reply:
I'd like to understand the larger context around this class before making too many meaningful changes.

Reply:
Yeah, agreed. I can see a few race conditions in this code but we'd need to sit down to see what the outcome of such a thing is